batchglm/models/base_glm/model.py

```diff
@@ -221,8 +277,148 @@ def get(self, key: Union[str, Iterable]) -> Union[Any, Dict[str, Any]]:
+
+    def param_bounds(self, dtype) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+
+        dtype = np.dtype(dtype)
+        # dmin = np.finfo(dtype).min
+        dmax = np.finfo(dtype).max
+        dtype = dtype.type
+        sf = dtype(pkg_constants.ACCURACY_MARGIN_RELATIVE_TO_LIMIT)
+
+        return self.bounds(sf, dmax, dtype)
+
+    @abc.abstractmethod
+    def bounds(self, sf, dmax, dtype) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+        pass
+
+    # simulator:
+
+    @abc.abstractmethod
+    def rand_fn_ave(self) -> Optional[Callable]:
+        pass
+
+    @abc.abstractmethod
+    def rand_fn(self) -> Optional[Callable]:
+        pass
+
+    @abc.abstractmethod
+    def rand_fn_loc(self) -> Optional[Callable]:
+        pass
+
+    @abc.abstractmethod
+    def rand_fn_scale(self) -> Optional[Callable]:
+        pass
+
+    def generate_params(
+        self, n_vars: int, rand_fn_ave=None, rand_fn=None, rand_fn_loc=None, rand_fn_scale=None, **kwargs
+    ):
+        """
+        Generate all necessary parameters. TODO: make this documentation better!!!
+
+        :param rand_fn_ave: function which generates random numbers for intercept.
+            Takes one location parameter of intercept distribution across features.
+        :param rand_fn: random function taking one argument `shape`.
+        :param rand_fn_loc: random function taking one argument `shape`.
+            If not provided, will use `rand_fn` instead.
+            This function generates location model parameters in inverse linker space,
+            ie. these parameter will be log transformed if a log linker function is used!
+            Values below 1e-08 will be set to 1e-08 to map them into the positive support.
+        :param rand_fn_scale: random function taking one argument `shape`.
+            If not provided, will use `rand_fn` instead.
+            This function generates scale model parameters in inverse linker space,
+            ie. these parameter will be log transformed if a log linker function is used!
+            Values below 1e-08 will be set to 1e-08 to map them into the positive support.
+        """
+
+        if rand_fn_ave is None:
+            rand_fn_ave = self.rand_fn_ave
+            if rand_fn_ave is None:
+                raise ValueError("rand_fn_ave must not be None!")
+        if rand_fn is None:
+            rand_fn = self.rand_fn
+        if rand_fn_loc is None:
+            rand_fn_loc = self.rand_fn_loc
+        if rand_fn_scale is None:
+            rand_fn_scale = self.rand_fn_scale
+        if rand_fn is None and rand_fn_loc is None:
+            raise ValueError("rand_fn and rand_fn_loc must not be both None!")
+        if rand_fn is None and rand_fn_scale is None:
+            raise ValueError("rand_fn and rand_fn_scale must not be both None!")
+
+        if rand_fn_loc is None:
+            rand_fn_loc = rand_fn
+        if rand_fn_scale is None:
+            rand_fn_scale = rand_fn
```

Comment:
What exactly do you mean? np_clip_params, param_bounds and bounds is all required because these are distribution specific parameter clipping thresholds required to not run into numerical issues during fitting.
The rand functions are used to generate data for the simulation.

Reply:
ah ok, i missed that.  thanks!